# 1. power of index

In [1]:
import pandas as pd
import yfinance as yf

In [2]:
initialData  = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data", 
                     names = ['age','workclass','fnlwgt', 'education', 'education_num','marital_status','occupation','relationship','race','sex','capital_gain','capital_loss', 'hours_per_week', 'native_country','label'], 
                     index_col = False)
initialData.head(2)

age          workclass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   

        marital_status        occupation    relationship    race    sex  \
0        Never-married      Adm-clerical   Not-in-family   White   Male   
1   Married-civ-spouse   Exec-managerial         Husband   White   Male   

   capital_gain  capital_loss  hours_per_week  native_country   label  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K

In [3]:
# way to inspect the data
initialData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  label           32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


## picking random rows from dataframe

In [4]:
# the way to access the rows is by using loc => labels/name , iloc => position number
# getting 10000 records randomly and then sorting them by index values
data = initialData.sample(10000, random_state=100).sort_index(axis=0)
data

age          workclass  fnlwgt      education  education_num  \
3       53            Private  234721           11th              7   
4       28            Private  338409      Bachelors             13   
6       49            Private  160187            9th              5   
8       31            Private   45781        Masters             14   
11      30          State-gov  141297      Bachelors             13   
...    ...                ...     ...            ...            ...   
32542   72                  ?  129912        HS-grad              9   
32543   45          Local-gov  119199     Assoc-acdm             12   
32548   65   Self-emp-not-inc   99359    Prof-school             15   
32550   43   Self-emp-not-inc   27242   Some-college             10   
32556   27            Private  257302     Assoc-acdm             12   

               marital_status          occupation    relationship  \
3          Married-civ-spouse   Handlers-cleaners         Husband   
4          Married-civ-spouse      Prof-specialty            Wife   
6       Married-spouse-absent       Other-service   Not-in-family   
8               Never-married      Prof-specialty   Not-in-family   
11         Married-civ-spouse      Prof-specialty         Husband   
...                       ...                 ...             ...   
32542      Married-civ-spouse                   ?         Husband   
32543                Divorced      Prof-specialty       Unmarried   
32548           Never-married      Prof-specialty   Not-in-family   
32550      Married-civ-spouse        Craft-repair         Husband   
32556      Married-civ-spouse        Tech-support            Wife   

                      race      sex  capital_gain  capital_loss  \
3                    Black     Male             0             0   
4                    Black   Female             0             0   
6                    Black   Female             0             0   
8                    White   Female         14084             0   
11      Asian-Pac-Islander     Male             0             0   
...                    ...      ...           ...           ...   
32542                White     Male             0             0   
32543                White   Female             0             0   
32548                White     Male          1086             0   
32550                White     Male             0             0   
32556                White   Female             0             0   

       hours_per_week  native_country   label  
3                  40   United-States   <=50K  
4                  40            Cuba   <=50K  
6                  16         Jamaica   <=50K  
8                  50   United-States    >50K  
11                 40           India    >50K  
...               ...             ...     ...  
32542              25   United-States   <=50K  
32543              48   United-States   <=50K  
32548              60   United-States   <=50K  
32550              50   United-States   <=50K  
32556              38   United-States   <=50K  

[10000 rows x 15 columns]

## renaming index column

In [5]:
# renaming index
print(data.index.name)
data.index.name = 'index_data'
print(data.index.name)

None
index_data


In [6]:
data.head(5)

age   workclass  fnlwgt   education  education_num  \
index_data                                                       
3            53     Private  234721        11th              7   
4            28     Private  338409   Bachelors             13   
6            49     Private  160187         9th              5   
8            31     Private   45781     Masters             14   
11           30   State-gov  141297   Bachelors             13   

                    marital_status          occupation    relationship  \
index_data                                                               
3               Married-civ-spouse   Handlers-cleaners         Husband   
4               Married-civ-spouse      Prof-specialty            Wife   
6            Married-spouse-absent       Other-service   Not-in-family   
8                    Never-married      Prof-specialty   Not-in-family   
11              Married-civ-spouse      Prof-specialty         Husband   

                           race      sex  capital_gain  capital_loss  \
index_data                                                             
3                         Black     Male             0             0   
4                         Black   Female             0             0   
6                         Black   Female             0             0   
8                         White   Female         14084             0   
11           Asian-Pac-Islander     Male             0             0   

            hours_per_week  native_country   label  
index_data                                          
3                       40   United-States   <=50K  
4                       40            Cuba   <=50K  
6                       16         Jamaica   <=50K  
8                       50   United-States    >50K  
11                      40           India    >50K

In [7]:
# getting columnnames
data.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'label'],
      dtype='object')

## loc and iloc

In [8]:
# reading a column value filtering with index value
data.loc[3, ['age']]

age    53
Name: 3, dtype: object

In [9]:
# getting the entire row filtering with index value
data.loc[3]

age                                53
workclass                     Private
fnlwgt                         234721
education                        11th
education_num                       7
marital_status     Married-civ-spouse
occupation          Handlers-cleaners
relationship                  Husband
race                            Black
sex                              Male
capital_gain                        0
capital_loss                        0
hours_per_week                     40
native_country          United-States
label                           <=50K
Name: 3, dtype: object

In [10]:
# however usig iloc reads the data using index-position and not by index-value
data.iloc[3, 0]

np.int64(31)

In [11]:
# similarly to get the entire record
data.iloc[3]

age                            31
workclass                 Private
fnlwgt                      45781
education                 Masters
education_num                  14
marital_status      Never-married
occupation         Prof-specialty
relationship        Not-in-family
race                        White
sex                        Female
capital_gain                14084
capital_loss                    0
hours_per_week                 50
native_country      United-States
label                        >50K
Name: 8, dtype: object

In [12]:
data.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'label'],
      dtype='object')

In [13]:
# selecting only specific columns and getting rid of of the unwanted ones
cols = ['age','education', 'marital_status', 'race','sex','native_country','label']
data = data[cols]
data.head()

age   education          marital_status                 race  \
index_data                                                                 
3            53        11th      Married-civ-spouse                Black   
4            28   Bachelors      Married-civ-spouse                Black   
6            49         9th   Married-spouse-absent                Black   
8            31     Masters           Never-married                White   
11           30   Bachelors      Married-civ-spouse   Asian-Pac-Islander   

                sex  native_country   label  
index_data                                   
3              Male   United-States   <=50K  
4            Female            Cuba   <=50K  
6            Female         Jamaica   <=50K  
8            Female   United-States    >50K  
11             Male           India    >50K

## resetting index

In [14]:
#resetting index
data.reset_index(inplace=True)
data.head(5)

index_data  age   education          marital_status                 race  \
0           3   53        11th      Married-civ-spouse                Black   
1           4   28   Bachelors      Married-civ-spouse                Black   
2           6   49         9th   Married-spouse-absent                Black   
3           8   31     Masters           Never-married                White   
4          11   30   Bachelors      Married-civ-spouse   Asian-Pac-Islander   

       sex  native_country   label  
0     Male   United-States   <=50K  
1   Female            Cuba   <=50K  
2   Female         Jamaica   <=50K  
3   Female   United-States    >50K  
4     Male           India    >50K

In [15]:
data.index

RangeIndex(start=0, stop=10000, step=1)

In [16]:
#note that simply resetting the index , adds one more index column while the previous index column is left-untouched

## getting unique values of a given column

In [17]:
# for getting unique values in a particular column 
data.native_country.unique()

array([' United-States', ' Cuba', ' Jamaica', ' India', ' ?', ' Honduras',
       ' Puerto-Rico', ' Mexico', ' England', ' Italy', ' Thailand',
       ' Portugal', ' Philippines', ' France', ' Canada',
       ' Dominican-Republic', ' Germany', ' China', ' South', ' Japan',
       ' Iran', ' Outlying-US(Guam-USVI-etc)', ' Haiti',
       ' Trinadad&Tobago', ' Greece', ' El-Salvador', ' Taiwan',
       ' Poland', ' Guatemala', ' Vietnam', ' Cambodia', ' Ireland',
       ' Laos', ' Columbia', ' Yugoslavia', ' Nicaragua', ' Ecuador',
       ' Peru', ' Hungary', ' Holand-Netherlands', ' Hong', ' Scotland'],
      dtype=object)

In [18]:
indiaData = data[data.native_country  == ' India']
indiaData

index_data  age      education          marital_status  \
4             11   30      Bachelors      Married-civ-spouse   
312         1029   48        Masters   Married-spouse-absent   
637         2130   28           11th               Separated   
902         2989   44        Masters      Married-civ-spouse   
1365        4480   30      Bachelors      Married-civ-spouse   
1379        4523   57      Bachelors      Married-civ-spouse   
1705        5629   46   Some-college      Married-civ-spouse   
2119        6964   29      Doctorate      Married-civ-spouse   
2388        7844   35        Masters      Married-civ-spouse   
2467        8124   36    Prof-school           Never-married   
2796        9128   18        HS-grad           Never-married   
3016        9793   45        Masters      Married-civ-spouse   
3063        9939   43    Prof-school      Married-civ-spouse   
3265       10590   35    Prof-school           Never-married   
3285       10661   59    Prof-school      Married-civ-spouse   
3655       11835   38      Bachelors   Married-spouse-absent   
3811       12368   24   Some-college           Never-married   
4171       13551   40      Doctorate      Married-civ-spouse   
4515       14685   36        Masters                 Widowed   
5192       16923   41      Assoc-voc      Married-civ-spouse   
5471       17834   29        HS-grad   Married-spouse-absent   
6122       19926   27    Prof-school           Never-married   
6133       19959   45            9th      Married-civ-spouse   
6482       21128   30        Masters      Married-civ-spouse   
7391       24154   46      Bachelors      Married-civ-spouse   
7891       25739   35        Masters      Married-civ-spouse   
8069       26356   34        Masters      Married-civ-spouse   
8355       27244   33        Masters      Married-civ-spouse   
8551       27889   17           11th           Never-married   
8580       27974   33      Bachelors      Married-civ-spouse   
8687       28264   51    Prof-school      Married-civ-spouse   
8742       28433   42        Masters      Married-civ-spouse   
8751       28452   53        Masters      Married-civ-spouse   
8820       28693   32           11th           Never-married   
8853       28798   30        Masters                Divorced   
9513       30871   26        Masters           Never-married   
9644       31327   38      Bachelors      Married-civ-spouse   

                     race      sex native_country   label  
4      Asian-Pac-Islander     Male          India    >50K  
312    Asian-Pac-Islander     Male          India   <=50K  
637    Asian-Pac-Islander   Female          India   <=50K  
902    Asian-Pac-Islander     Male          India   <=50K  
1365   Asian-Pac-Islander     Male          India   <=50K  
1379   Asian-Pac-Islander     Male          India   <=50K  
1705   Asian-Pac-Islander     Male          India   <=50K  
2119   Asian-Pac-Islander     Male          India   <=50K  
2388                White     Male          India   <=50K  
2467                Other     Male          India    >50K  
2796   Asian-Pac-Islander   Female          India   <=50K  
3016                White     Male          India   <=50K  
3063   Asian-Pac-Islander     Male          India    >50K  
3265   Asian-Pac-Islander     Male          India    >50K  
3285   Asian-Pac-Islander     Male          India    >50K  
3655                Other     Male          India   <=50K  
3811                White     Male          India   <=50K  
4171   Asian-Pac-Islander     Male          India    >50K  
4515   Asian-Pac-Islander   Female          India   <=50K  
5192   Asian-Pac-Islander     Male          India    >50K  
5471   Asian-Pac-Islander     Male          India    >50K  
6122   Asian-Pac-Islander     Male          India   <=50K  
6133   Asian-Pac-Islander     Male          India   <=50K  
6482   Asian-Pac-Islander     Male          India    >50K  
7391                White     Male          India    >50K  
7891   Asian-

## crosstab

In [19]:
# using cross tab
cross = pd.crosstab(indiaData.sex, indiaData.label)
cross

label   <=50K  >50K
sex                
Female      4     0
Male       16    17

In [20]:
cross.index

Index([' Female', ' Male'], dtype='object', name='sex')

In [21]:
# to filter only female record. thus we have a requriement to filter by index value hence we'l be using loc
cross.loc[' Female']

label
<=50K    4
>50K     0
Name:  Female, dtype: int64

In [22]:
# doing filter on the actual dataframe to verify the count in the crosstab
indiaData_gt_50  = indiaData[(indiaData.label ==' <=50K') & (indiaData.sex==' Female')]
len(indiaData_gt_50)

4

# 2. handling missing data

In [23]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='1y')
tickerData = tickerData['Close']

YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  2 of 2 completed


In [24]:
# gives me all the index values
tickerData.index

DatetimeIndex(['2024-03-01', '2024-03-04', '2024-03-05', '2024-03-06',
               '2024-03-07', '2024-03-08', '2024-03-11', '2024-03-12',
               '2024-03-13', '2024-03-14',
               ...
               '2025-02-14', '2025-02-18', '2025-02-19', '2025-02-20',
               '2025-02-21', '2025-02-24', '2025-02-25', '2025-02-26',
               '2025-02-27', '2025-02-28'],
              dtype='datetime64[ns]', name='Date', length=250, freq=None)

## extracting date related info from index

In [25]:
#to see only the day
tickerData.index.day

Index([ 1,  4,  5,  6,  7,  8, 11, 12, 13, 14,
       ...
       14, 18, 19, 20, 21, 24, 25, 26, 27, 28],
      dtype='int32', name='Date', length=250)

In [26]:
#to see only year
tickerData.index.year

Index([2024, 2024, 2024, 2024, 2024, 2024, 2024, 2024, 2024, 2024,
       ...
       2025, 2025, 2025, 2025, 2025, 2025, 2025, 2025, 2025, 2025],
      dtype='int32', name='Date', length=250)

In [27]:
#optionally we can add additinal columns as shown below 
tickerData['day_name'] = tickerData.index.day_name()
tickerData['isEndOfMonth'] = tickerData.index.is_month_end
tickerData.head(5)

Ticker            AAPL        TSLA   day_name  isEndOfMonth
Date                                                       
2024-03-01  178.815659  202.639999     Friday         False
2024-03-04  174.277100  188.139999     Monday         False
2024-03-05  169.320496  180.740005    Tuesday         False
2024-03-06  168.325195  176.539993  Wednesday         False
2024-03-07  168.205750  178.649994   Thursday         False

In [28]:
# we can see all the attributes using as below 
help(tickerData.index)

Help on DatetimeIndex in module pandas.core.indexes.datetimes object:

class DatetimeIndex(pandas.core.indexes.datetimelike.DatetimeTimedeltaMixin)
 |  DatetimeIndex(data=None, freq: 'Frequency | lib.NoDefault' = <no_default>, tz=<no_default>, normalize: 'bool | lib.NoDefault' = <no_default>, closed=<no_default>, ambiguous: 'TimeAmbiguous' = 'raise', dayfirst: 'bool' = False, yearfirst: 'bool' = False, dtype: 'Dtype | None' = None, copy: 'bool' = False, name: 'Hashable | None' = None) -> 'Self'
 |
 |  Immutable ndarray-like of datetime64 data.
 |
 |  Represented internally as int64, and which can be boxed to Timestamp objects
 |  that are subclasses of datetime and carry metadata.
 |
 |  .. versionchanged:: 2.0.0
 |      The various numeric date/time attributes (:attr:`~DatetimeIndex.day`,
 |      :attr:`~DatetimeIndex.month`, :attr:`~DatetimeIndex.year` etc.) now have dtype
 |      ``int32``. Previously they had dtype ``int64``.
 |
 |  Parameters
 |  ----------
 |  data : array-like (

## creating a dataframe with my daterange

In [29]:
# if we notice the intial dataframe we have data from 2024-02-22 to 2025-02-21
# lets say we want to add empty dates i.e. the holday to the dataframe this is the way to go about it. 

#here we are creating a data frame by including all the dates for a slightly expanded date range
all_dates = pd.date_range(start='2024-02-18', end = '2025-02-28', freq='1d')
all_dates

DatetimeIndex(['2024-02-18', '2024-02-19', '2024-02-20', '2024-02-21',
               '2024-02-22', '2024-02-23', '2024-02-24', '2024-02-25',
               '2024-02-26', '2024-02-27',
               ...
               '2025-02-19', '2025-02-20', '2025-02-21', '2025-02-22',
               '2025-02-23', '2025-02-24', '2025-02-25', '2025-02-26',
               '2025-02-27', '2025-02-28'],
              dtype='datetime64[ns]', length=377, freq='D')

In [30]:
# now to addtional days can added to the initial dataframe as shown below:
tickerData = tickerData.reindex(all_dates)
tickerData.head(8)

Ticker      AAPL  TSLA day_name isEndOfMonth
2024-02-18   NaN   NaN      NaN          NaN
2024-02-19   NaN   NaN      NaN          NaN
2024-02-20   NaN   NaN      NaN          NaN
2024-02-21   NaN   NaN      NaN          NaN
2024-02-22   NaN   NaN      NaN          NaN
2024-02-23   NaN   NaN      NaN          NaN
2024-02-24   NaN   NaN      NaN          NaN
2024-02-25   NaN   NaN      NaN          NaN

thus the non-working days also included now. and they have NaN for the other columns as value

In [31]:
# now if want the day_name, isEndOfMonth to be shown for the newly added records. then those need to be revalueated as
tickerData['day_name'] = tickerData.index.day_name()
tickerData['isEndOfMonth'] = tickerData.index.is_month_end
tickerData.head(10)

Ticker      AAPL  TSLA   day_name  isEndOfMonth
2024-02-18   NaN   NaN     Sunday         False
2024-02-19   NaN   NaN     Monday         False
2024-02-20   NaN   NaN    Tuesday         False
2024-02-21   NaN   NaN  Wednesday         False
2024-02-22   NaN   NaN   Thursday         False
2024-02-23   NaN   NaN     Friday         False
2024-02-24   NaN   NaN   Saturday         False
2024-02-25   NaN   NaN     Sunday         False
2024-02-26   NaN   NaN     Monday         False
2024-02-27   NaN   NaN    Tuesday         False

## filling missing data

### bfill - backfill
it takes the first available value inside a columne and then it propagates it upwards

In [32]:
#tickerData.fillna(method='bfill')
tickerData.bfill(axis=0, inplace=True)
tickerData.head(10)

Ticker            AAPL        TSLA   day_name  isEndOfMonth
2024-02-18  178.815659  202.639999     Sunday         False
2024-02-19  178.815659  202.639999     Monday         False
2024-02-20  178.815659  202.639999    Tuesday         False
2024-02-21  178.815659  202.639999  Wednesday         False
2024-02-22  178.815659  202.639999   Thursday         False
2024-02-23  178.815659  202.639999     Friday         False
2024-02-24  178.815659  202.639999   Saturday         False
2024-02-25  178.815659  202.639999     Sunday         False
2024-02-26  178.815659  202.639999     Monday         False
2024-02-27  178.815659  202.639999    Tuesday         False

### ffill -forward fill

In [33]:
tickerData.tail(10)

Ticker            AAPL        TSLA   day_name  isEndOfMonth
2025-02-19  244.869995  360.559998  Wednesday         False
2025-02-20  245.830002  354.399994   Thursday         False
2025-02-21  245.550003  337.799988     Friday         False
2025-02-22  247.100006  330.529999   Saturday         False
2025-02-23  247.100006  330.529999     Sunday         False
2025-02-24  247.100006  330.529999     Monday         False
2025-02-25  247.039993  302.799988    Tuesday         False
2025-02-26  240.360001  290.799988  Wednesday         False
2025-02-27  237.300003  281.950012   Thursday         False
2025-02-28  241.839996  292.980011     Friday          True

In [34]:
# similarly we can do the opposite using forward fill
tickerData.ffill(axis=0, inplace=True)
tickerData.tail(10)

Ticker            AAPL        TSLA   day_name  isEndOfMonth
2025-02-19  244.869995  360.559998  Wednesday         False
2025-02-20  245.830002  354.399994   Thursday         False
2025-02-21  245.550003  337.799988     Friday         False
2025-02-22  247.100006  330.529999   Saturday         False
2025-02-23  247.100006  330.529999     Sunday         False
2025-02-24  247.100006  330.529999     Monday         False
2025-02-25  247.039993  302.799988    Tuesday         False
2025-02-26  240.360001  290.799988  Wednesday         False
2025-02-27  237.300003  281.950012   Thursday         False
2025-02-28  241.839996  292.980011     Friday          True

# 3.Resampling of time-series data

In [35]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y').Close
tickerData .head()

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL        TSLA
Date                              
2022-03-01  160.597763  288.123322
2022-03-02  163.904190  293.296661
2022-03-03  163.579437  279.763336
2022-03-04  160.568207  279.429993
2022-03-07  156.759949  268.193329

## aggregation methods - agg()

getting the first date for every year . this is where the magic of resampling comes in 

In [36]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
tickerData.head()

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL        TSLA
Date                              
2022-03-01  160.597763  288.123322
2022-03-02  163.904190  293.296661
2022-03-03  163.579437  279.763336
2022-03-04  160.568207  279.429993
2022-03-07  156.759949  268.193329

In [37]:
# Cast the index to a DatetimeIndex
tickerData.index = pd.DatetimeIndex(tickerData.index)

# Resample the data
tickerData_resampled = tickerData.resample("YS").first() 
#YS is for year-start. similarly YE - is for  . as for the aggregation we can apply various aggregate methods like first(), last(), min(), max(), avg()
tickerData_resampled

Ticker            AAPL        TSLA
Date                              
2022-01-01  160.597763  288.123322
2023-01-01  123.632538  108.099998
2024-01-01  184.532089  248.419998
2025-01-01  243.582199  379.279999

## getting the max annual price for every year

In [38]:
tickers = ['AAP', 'TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
max_price_by_year = tickerData.resample('YE').max()
max_price_by_year

[*********************100%***********************]  2 of 2 completed


Ticker             AAP        TSLA
Date                              
2022-12-31  212.057632  381.816681
2023-12-31  150.019730  293.339996
2024-12-31   84.781960  479.859985
2025-12-31   49.790001  428.220001

## getting minimum close prices per year

In [39]:
min_price_by_year = tickerData.resample('YE').min()
min_price_by_year

Ticker             AAP        TSLA
Date                              
2022-12-31  133.318970  109.099998
2023-12-31   47.125027  108.099998
2024-12-31   35.494030  142.050003
2025-12-31   36.900002  281.950012

In [40]:
# getting minimum close prices per quarter
min_price_by_quarter = tickerData.resample('QE').min()
min_price_by_quarter

Ticker             AAP        TSLA
Date                              
2022-03-31  183.888367  255.456665
2022-06-30  155.219193  209.386673
2022-09-30  147.977905  227.263336
2022-12-31  133.318970  109.099998
2023-03-31  106.010498  108.099998
2023-06-30   61.576393  153.750000
2023-09-30   53.280716  215.490005
2023-12-31   47.125027  197.360001
2024-03-31   59.079647  162.500000
2024-06-30   60.792744  142.050003
2024-09-30   38.062298  191.759995
2024-12-31   35.494030  213.649994
2025-03-31   36.900002  281.950012

## multiple aggregations

In [41]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
tickerData.index = pd.DatetimeIndex(tickerData.index)

resampled_data = tickerData.resample('YE').agg(['min','max'])
resampled_data

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL                    TSLA            
                   min         max         min         max
Date                                                      
2022-12-31  124.591377  176.106476  109.099998  381.816681
2023-12-31  123.583099  196.927673  108.099998  293.339996
2024-12-31  164.224564  258.735504  142.050003  479.859985
2025-12-31  222.395477  247.100006  281.950012  428.220001

mixing aggregations . i.e fetching min and max for apple while fist and last for tesla

In [42]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
tickerData.index = pd.DatetimeIndex(tickerData.index)

# Resample by year and calculate the desired values
resampled_data = tickerData.resample('YE').agg({
    'AAPL': ['min', 'max'],
    'TSLA': ['first', 'last']
})

resampled_data

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL                    TSLA            
                   min         max       first        last
Date                                                      
2022-12-31  124.591377  176.106476  288.123322  123.180000
2023-12-31  123.583099  196.927673  108.099998  248.479996
2024-12-31  164.224564  258.735504  248.419998  403.839996
2025-12-31  222.395477  247.100006  379.279999  292.980011

In [43]:
resampled_data.index = resampled_data.index.year
resampled_data

Ticker        AAPL                    TSLA            
               min         max       first        last
Date                                                  
2022    124.591377  176.106476  288.123322  123.180000
2023    123.583099  196.927673  108.099998  248.479996
2024    164.224564  258.735504  248.419998  403.839996
2025    222.395477  247.100006  379.279999  292.980011

ohlc

In [44]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
tickerData.index = pd.DatetimeIndex(tickerData.index)
resampled_data = tickerData.resample('YE').ohlc()
resampled_data

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL                                            TSLA  \
                  open        high         low       close        open   
Date                                                                     
2022-12-31  160.597763  176.106476  124.591377  128.436646  288.123322   
2023-12-31  123.632538  196.927673  123.583099  191.380951  108.099998   
2024-12-31  184.532089  258.735504  164.224564  250.144974  248.419998   
2025-12-31  243.582199  247.100006  222.395477  241.839996  379.279999   

Ticker                                          
                  high         low       close  
Date                                            
2022-12-31  381.816681  109.099998  123.180000  
2023-12-31  293.339996  108.099998  248.479996  
2024-12-31  479.859985  142.050003  403.839996  
2025-12-31  428.220001  281.950012  292.980011

this means that the closing price of apple for the year 2022 :
* opened_with = 160.597763
* made a high of = 176.106461	
* made a low of = 124.591377	
* closed the year with = 128.436646

# 4.timezone travel with timeseries

In [45]:
#since we are interested in the timezone so not gettting daily data. rather fetching some interval data so that time is part of the index

In [46]:
tickers = "MSFT"
data = yf.download(tickers, period='7d', interval='30m')
data.head(2)

[*********************100%***********************]  1 of 1 completed


Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 14:30:00+00:00  413.890015  417.309906  413.565002  415.529999   
2025-02-20 15:00:00+00:00  416.109985  416.209991  412.540009  413.890015   

Price                       Volume  
Ticker                        MSFT  
Datetime                            
2025-02-20 14:30:00+00:00  4121360  
2025-02-20 15:00:00+00:00  1870450

## getting timezone info

In [47]:
# now this will give us the timezone of the index
data.index.tz

datetime.timezone.utc

## converting timezone

In [48]:
# now to change the timezone to easter standar time use below code.
#data.index = data.index.tz_convert('US/Eastern')

#am changing the times to ist and can be dont as shown below 
data.index = data.index.tz_convert('Asia/Kolkata')
data.head(2)



Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 20:00:00+05:30  413.890015  417.309906  413.565002  415.529999   
2025-02-20 20:30:00+05:30  416.109985  416.209991  412.540009  413.890015   

Price                       Volume  
Ticker                        MSFT  
Datetime                            
2025-02-20 20:00:00+05:30  4121360  
2025-02-20 20:30:00+05:30  1870450

In [49]:
data.index = data.index.tz_convert('Asia/Singapore')
data.head(2)

Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 22:30:00+08:00  413.890015  417.309906  413.565002  415.529999   
2025-02-20 23:00:00+08:00  416.109985  416.209991  412.540009  413.890015   

Price                       Volume  
Ticker                        MSFT  
Datetime                            
2025-02-20 22:30:00+08:00  4121360  
2025-02-20 23:00:00+08:00  1870450

In [50]:
data.index = data.index.tz_convert('America/New_York')
data.head(2)

Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 09:30:00-05:00  413.890015  417.309906  413.565002  415.529999   
2025-02-20 10:00:00-05:00  416.109985  416.209991  412.540009  413.890015   

Price                       Volume  
Ticker                        MSFT  
Datetime                            
2025-02-20 09:30:00-05:00  4121360  
2025-02-20 10:00:00-05:00  1870450

## creating a dataframe from another dataframe while changing the timezone of the index

earlier we were modifying the timezone of the index of the dataframe from one to the another timezone

In [51]:
tickers = "MSFT"
data = yf.download(tickers, period='7d', interval='30m')

# Create a new DataFrame with the index converted to IST
data_ist = pd.DataFrame(data.values, index=data.index.tz_convert('Asia/Kolkata'), columns=data.columns)
data_ist.head()

[*********************100%***********************]  1 of 1 completed


Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 20:00:00+05:30  413.890015  417.309906  413.565002  415.529999   
2025-02-20 20:30:00+05:30  416.109985  416.209991  412.540009  413.890015   
2025-02-20 21:00:00+05:30  415.984985  416.950012  415.640015  416.089996   
2025-02-20 21:30:00+05:30  417.920105  419.309998  416.049988  416.209991   
2025-02-20 22:00:00+05:30  416.589996  418.100006  415.700012  418.049988   

Price                         Volume  
Ticker                          MSFT  
Datetime                              
2025-02-20 20:00:00+05:30  4121360.0  
2025-02-20 20:30:00+05:30  1870450.0  
2025-02-20 21:00:00+05:30  1239396.0  
2025-02-20 21:30:00+05:30  2507475.0  
2025-02-20 22:00:00+05:30  1069458.0

In [52]:
# lets create one more dataframe form the same data but with a different timezone 
# Create a new DataFrame with the index converted to IST
data_ny = pd.DataFrame(data.values, index=data.index.tz_convert('America/New_York'), columns=data.columns)
data_ny.head()

Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 09:30:00-05:00  413.890015  417.309906  413.565002  415.529999   
2025-02-20 10:00:00-05:00  416.109985  416.209991  412.540009  413.890015   
2025-02-20 10:30:00-05:00  415.984985  416.950012  415.640015  416.089996   
2025-02-20 11:00:00-05:00  417.920105  419.309998  416.049988  416.209991   
2025-02-20 11:30:00-05:00  416.589996  418.100006  415.700012  418.049988   

Price                         Volume  
Ticker                          MSFT  
Datetime                              
2025-02-20 09:30:00-05:00  4121360.0  
2025-02-20 10:00:00-05:00  1870450.0  
2025-02-20 10:30:00-05:00  1239396.0  
2025-02-20 11:00:00-05:00  2507475.0  
2025-02-20 11:30:00-05:00  1069458.0

## combining dataframes with multiple timezones

In [53]:
alltimes = pd.concat([data_ny, data_ist], axis=1) # this will combine that dataframe and display the values side by side for the same index value
alltimes

Price                           Close        High         Low        Open  \
Ticker                           MSFT        MSFT        MSFT        MSFT   
Datetime                                                                    
2025-02-20 14:30:00+00:00  413.890015  417.309906  413.565002  415.529999   
2025-02-20 15:00:00+00:00  416.109985  416.209991  412.540009  413.890015   
2025-02-20 15:30:00+00:00  415.984985  416.950012  415.640015  416.089996   
2025-02-20 16:00:00+00:00  417.920105  419.309998  416.049988  416.209991   
2025-02-20 16:30:00+00:00  416.589996  418.100006  415.700012  418.049988   
...                               ...         ...         ...         ...   
2025-02-28 18:30:00+00:00  389.690002  392.048096  389.690002  391.170013   
2025-02-28 19:00:00+00:00  390.859985  391.250000  389.579987  389.654999   
2025-02-28 19:30:00+00:00  392.390015  392.500000  390.567596  390.855011   
2025-02-28 20:00:00+00:00  393.070007  393.119904  391.769989  392.410004   
2025-02-28 20:30:00+00:00  397.142487  397.630005  391.570007  393.070007   

Price                         Volume       Close        High         Low  \
Ticker                          MSFT        MSFT        MSFT        MSFT   
Datetime                                                                   
2025-02-20 14:30:00+00:00  4121360.0  413.890015  417.309906  413.565002   
2025-02-20 15:00:00+00:00  1870450.0  416.109985  416.209991  412.540009   
2025-02-20 15:30:00+00:00  1239396.0  415.984985  416.950012  415.640015   
2025-02-20 16:00:00+00:00  2507475.0  417.920105  419.309998  416.049988   
2025-02-20 16:30:00+00:00  1069458.0  416.589996  418.100006  415.700012   
...                              ...         ...         ...         ...   
2025-02-28 18:30:00+00:00   738267.0  389.690002  392.048096  389.690002   
2025-02-28 19:00:00+00:00   992909.0  390.859985  391.250000  389.579987   
2025-02-28 19:30:00+00:00  1393041.0  392.390015  392.500000  390.567596   
2025-02-28 20:00:00+00:00   882549.0  393.070007  393.119904  391.769989   
2025-02-28 20:30:00+00:00  3857564.0  397.142487  397.630005  391.570007   

Price                            Open     Volume  
Ticker                           MSFT       MSFT  
Datetime                                          
2025-02-20 14:30:00+00:00  415.529999  4121360.0  
2025-02-20 15:00:00+00:00  413.890015  1870450.0  
2025-02-20 15:30:00+00:00  416.089996  1239396.0  
2025-02-20 16:00:00+00:00  416.209991  2507475.0  
2025-02-20 16:30:00+00:00  418.049988  1069458.0  
...                               ...        ...  
2025-02-28 18:30:00+00:00  391.170013   738267.0  
2025-02-28 19:00:00+00:00  389.654999   992909.0  
2025-02-28 19:30:00+00:00  390.855011  1393041.0  
2025-02-28 20:00:00+00:00  392.410004   882549.0  
2025-02-28 20:30:00+00:00  393.070007  3857564.0  

[91 rows x 10 columns]

## shifting dates    

In [54]:
ts = pd.Timestamp('2025-01-01')
ts

Timestamp('2025-01-01 00:00:00')

In [55]:
# adding a day to it
print(ts +pd.DateOffset(days=1))

2025-01-02 00:00:00


In [56]:
# adding a month to it
print(ts +pd.DateOffset(months=1))

2025-02-01 00:00:00


In [57]:
# adding an year
print(ts +pd.DateOffset(years=1))

2026-01-01 00:00:00


In [58]:
tickers = ['AAPL','TSLA']
tickerData = yf.download(tickers, period='3y')['Close']
tickerData

[*********************100%***********************]  2 of 2 completed


Ticker            AAPL        TSLA
Date                              
2022-03-01  160.597763  288.123322
2022-03-02  163.904190  293.296661
2022-03-03  163.579437  279.763336
2022-03-04  160.568207  279.429993
2022-03-07  156.759949  268.193329
...                ...         ...
2025-02-24  247.100006  330.529999
2025-02-25  247.039993  302.799988
2025-02-26  240.360001  290.799988
2025-02-27  237.300003  281.950012
2025-02-28  241.839996  292.980011

[753 rows x 2 columns]

In [59]:
tickerData.index

DatetimeIndex(['2022-03-01', '2022-03-02', '2022-03-03', '2022-03-04',
               '2022-03-07', '2022-03-08', '2022-03-09', '2022-03-10',
               '2022-03-11', '2022-03-14',
               ...
               '2025-02-14', '2025-02-18', '2025-02-19', '2025-02-20',
               '2025-02-21', '2025-02-24', '2025-02-25', '2025-02-26',
               '2025-02-27', '2025-02-28'],
              dtype='datetime64[ns]', name='Date', length=753, freq=None)

In [60]:
# now lets push the days by 1
print("index values after adding 1 day to it.",tickerData.index + pd.DateOffset(days=1))


# thus here we have added 1 day to every value in the index-series. however this doesnt change the actual index-values of the dataframe
print("printing actual index-values.",tickerData.index)

index values after adding 1 day to it. DatetimeIndex(['2022-03-02', '2022-03-03', '2022-03-04', '2022-03-05',
               '2022-03-08', '2022-03-09', '2022-03-10', '2022-03-11',
               '2022-03-12', '2022-03-15',
               ...
               '2025-02-15', '2025-02-19', '2025-02-20', '2025-02-21',
               '2025-02-22', '2025-02-25', '2025-02-26', '2025-02-27',
               '2025-02-28', '2025-03-01'],
              dtype='datetime64[ns]', name='Date', length=753, freq=None)
printing actual index-values. DatetimeIndex(['2022-03-01', '2022-03-02', '2022-03-03', '2022-03-04',
               '2022-03-07', '2022-03-08', '2022-03-09', '2022-03-10',
               '2022-03-11', '2022-03-14',
               ...
               '2025-02-14', '2025-02-18', '2025-02-19', '2025-02-20',
               '2025-02-21', '2025-02-24', '2025-02-25', '2025-02-26',
               '2025-02-27', '2025-02-28'],
              dtype='datetime64[ns]', name='Date', length=753, freq=None)


In [61]:
# however if want to change the actual index-values in the dataframe then that can be done by assiging the calculated value back to the index
tickerData.index = tickerData.index + pd.DateOffset(days=1)
tickerData.head(2)


Ticker            AAPL        TSLA
Date                              
2022-03-02  160.597763  288.123322
2022-03-03  163.904190  293.296661

# 5. finding the largest and smallest data points

In [62]:
tickers = 'AAPL'
tickerData = yf.download(tickers, period='3y')
tickerData

[*********************100%***********************]  1 of 1 completed


Price            Close        High         Low        Open    Volume
Ticker            AAPL        AAPL        AAPL        AAPL      AAPL
Date                                                                
2022-03-01  160.597763  163.943559  159.387380  162.073845  83474400
2022-03-02  163.904190  164.691437  160.351751  161.768793  79724800
2022-03-03  163.579437  166.216712  162.910287  165.783726  76678400
2022-03-04  160.568207  162.910262  159.515276  161.867166  83737200
2022-03-07  156.759949  162.388744  156.504085  160.755209  96418800
...                ...         ...         ...         ...       ...
2025-02-24  247.100006  248.860001  244.419998  244.929993  51326400
2025-02-25  247.039993  250.000000  244.910004  248.000000  48013300
2025-02-26  240.360001  244.979996  239.130005  244.330002  44433600
2025-02-27  237.300003  242.460007  237.059998  239.410004  41153600
2025-02-28  241.839996  242.089996  230.199997  236.949997  56796200

[753 rows x 5 columns]

## finding percentage change

In [63]:
tickerData["daily_return_percent"] = tickerData['Close'].pct_change()

In [64]:
tickerData.head(5)

Price            Close        High         Low        Open    Volume  \
Ticker            AAPL        AAPL        AAPL        AAPL      AAPL   
Date                                                                   
2022-03-01  160.597763  163.943559  159.387380  162.073845  83474400   
2022-03-02  163.904190  164.691437  160.351751  161.768793  79724800   
2022-03-03  163.579437  166.216712  162.910287  165.783726  76678400   
2022-03-04  160.568207  162.910262  159.515276  161.867166  83737200   
2022-03-07  156.759949  162.388744  156.504085  160.755209  96418800   

Price      daily_return_percent  
Ticker                           
Date                             
2022-03-01                  NaN  
2022-03-02             0.020588  
2022-03-03            -0.001981  
2022-03-04            -0.018408  
2022-03-07            -0.023717

## finding top 10 days of volume

even before finding the top 10 days with the highest volume first lets simplify the dataframe by removing column-level-1.
as currently its a multi-indexed dataframe so removing one level to make it simple

In [65]:
tickerData.columns = tickerData.columns.droplevel(1)
tickerData

Price            Close        High         Low        Open    Volume  \
Date                                                                   
2022-03-01  160.597763  163.943559  159.387380  162.073845  83474400   
2022-03-02  163.904190  164.691437  160.351751  161.768793  79724800   
2022-03-03  163.579437  166.216712  162.910287  165.783726  76678400   
2022-03-04  160.568207  162.910262  159.515276  161.867166  83737200   
2022-03-07  156.759949  162.388744  156.504085  160.755209  96418800   
...                ...         ...         ...         ...       ...   
2025-02-24  247.100006  248.860001  244.419998  244.929993  51326400   
2025-02-25  247.039993  250.000000  244.910004  248.000000  48013300   
2025-02-26  240.360001  244.979996  239.130005  244.330002  44433600   
2025-02-27  237.300003  242.460007  237.059998  239.410004  41153600   
2025-02-28  241.839996  242.089996  230.199997  236.949997  56796200   

Price       daily_return_percent  
Date                              
2022-03-01                   NaN  
2022-03-02              0.020588  
2022-03-03             -0.001981  
2022-03-04             -0.018408  
2022-03-07             -0.023717  
...                          ...  
2025-02-24              0.006312  
2025-02-25             -0.000243  
2025-02-26             -0.027040  
2025-02-27             -0.012731  
2025-02-28              0.019132  

[753 rows x 6 columns]

In [66]:
top_10_volume_days = tickerData.nlargest(10, 'Volume')
top_10_volume_days

Price            Close        High         Low        Open     Volume  \
Date                                                                    
2024-09-20  227.698853  232.578113  227.120124  229.464970  318679900   
2024-06-21  206.794983  211.180238  206.416251  209.685263  246421400   
2024-06-12  212.356308  219.462415  206.206962  206.675389  198134300   
2022-05-12  140.492966  144.080187  136.787489  140.699927  182602000   
2024-06-11  206.456116  206.466092  192.981414  193.001336  172373300   
2022-10-28  153.695084  155.431969  145.879078  146.254078  164762400   
2024-05-03  182.518173  186.121155  181.801556  185.772794  163224100   
2022-09-16  148.721222  149.362696  146.421814  149.224535  162278800   
2022-12-16  132.964035  136.067946  132.193001  135.118988  160156900   
2023-02-03  152.724258  155.571162  146.130922  146.328620  154357300   

Price       daily_return_percent  
Date                              
2024-09-20             -0.002927  
2024-06-21             -0.010444  
2024-06-12              0.028578  
2022-05-12             -0.026894  
2024-06-11              0.072649  
2022-10-28              0.075553  
2024-05-03              0.059816  
2022-09-16             -0.010960  
2022-12-16             -0.014579  
2023-02-03              0.024400

## finding days with the highest percentage returns

In [67]:
top_10_returns_daily = tickerData.nlargest(10, 'daily_return_percent')
top_10_returns_daily

Price            Close        High         Low        Open     Volume  \
Date                                                                    
2022-11-10  145.181961  145.181961  137.896672  139.616679  118854000   
2022-10-28  153.695084  155.431969  145.879078  146.254078  164762400   
2024-06-11  206.456116  206.466092  192.981414  193.001336  172373300   
2024-05-03  182.518173  186.121155  181.801556  185.772794  163224100   
2022-11-30  146.328644  147.010716  138.934618  139.774840  111380900   
2023-05-05  171.837067  172.559774  169.055109  169.272914  113316400   
2022-04-28  161.030731  161.896704  156.395826  156.710731  130216800   
2024-04-11  174.217377  174.635416  167.369720  167.548867   91070300   
2022-05-04  163.372757  163.825414  156.720538  157.124004  108256500   
2022-05-27  147.470337  147.509750  143.153839  143.281959   90978500   

Price       daily_return_percent  
Date                              
2022-11-10              0.088975  
2022-10-28              0.075553  
2024-06-11              0.072649  
2024-05-03              0.059816  
2022-11-30              0.048594  
2023-05-05              0.046927  
2022-04-28              0.045155  
2024-04-11              0.043271  
2022-05-04              0.041008  
2022-05-27              0.040757